File: src/aoc2015/jav/Day16.java
package aoc2015.jav;

import java.io.IOException;
import java.nio.file.*;
import java.util.*;
import java.util.stream.Collectors;

class Day16 {

  private static int solve(List<String> s, Map<String, Integer> properties, boolean partTwo) {
    for (Map<String, Integer> aunt : parseAunts(s)) {
      if (aunt.entrySet().stream().allMatch(entry -> {
        if (entry.getKey().equals("Sue"))
          return true;
        if (partTwo && (entry.getKey().equals("cats:") || entry.getKey().equals("trees:"))) {
          return entry.getValue().compareTo(properties.get(entry.getKey())) > 0;
        } else if (partTwo && (entry.getKey().equals("pomeranians:") || entry.getKey().equals("goldfish:"))) {
          return entry.getValue().compareTo(properties.get(entry.getKey())) < 0;
        } else {
          return properties.get(entry.getKey()).equals(entry.getValue());
        }
      }))
        return aunt.get("Sue");
    }
    return -1;
  }

  private static List<Map<String, Integer>> parseAunts(List<String> list) {
    return list.stream().map(t -> {
      String[] line = t.split(" ");
      Map<String, Integer> a = new HashMap<>();
      for (int i = 0; i <= line.length - 2; i += 2) {
        a.put(line[i], Integer.valueOf(line[i + 1].replace(":", "").replace(",", "")));
      }
      return a;
    }).collect(Collectors.toList());
  }

  public static void main(String[] args) throws IOException {
    List<String> s = Files.readAllLines(Paths.get("./input/2015/Day16_input.txt"));
    Map<String, Integer> properties = new HashMap<>();
    properties.put("children:", 3);
    properties.put("cats:", 7);
    properties.put("samoyeds:", 2);
    properties.put("pomeranians:", 3);
    properties.put("akitas:", 0);
    properties.put("vizslas:", 0);
    properties.put("goldfish:", 5);
    properties.put("trees:", 3);
    properties.put("cars:", 2);
    properties.put("perfumes:", 1);
    System.out.println("Part One = " + solve(s, properties, false));
    System.out.println("Part Two = " + solve(s, properties, true));
  }
}

File: src/aoc2015/kot/Day16.kt
package aoc2015.kot

import itertools.intoPairs
import java.nio.file.Files
import java.nio.file.Paths

object Day16 {
  data class Sue(val num: Int, val stats: HashMap<String, Int>)

  fun parseAunts(lines: List<String>): List<Sue> {
    return lines.map {
      val attrs = hashMapOf<String, Int>()
      var num = 0
      for ((k, v) in it.split(" ").intoPairs()) {
        val key = k.filter { it != ':' && it != ',' }
        val value = Integer.valueOf(v?.filter { it != ':' && it != ',' })
        if (key == "Sue") num = value
        else attrs[key] = value
      }
      Sue(num, attrs)
    }
  }

  fun solve(aunts: List<Sue>, foundAttrs: HashMap<String, Int>, filter: (String, Int, Int) -> Boolean): Int {
    return aunts.first { it.stats.all { (k, v) -> filter(k, foundAttrs[k]!!, v) } }.num
  }
}

fun main(args: Array<String>) {
  val s = Files.readAllLines(Paths.get("./input/2015/Day16_input.txt"))
  val sues = Day16.parseAunts(s)
  val foundAttrs = hashMapOf("after" to 3, "cats" to 7, "samoyed" to 2, "pomeranians" to 3, "akitas" to 0, "vizslas" to 0, "goldfish" to 5, "trees" to 3, "cars" to 2, "perfumes" to 1)
  println("Part One = ${Day16.solve(sues, foundAttrs) { _: String, fv: Int, v: Int -> fv == v }}")

  fun filter(key: String, fv: Int, v: Int): Boolean {
    if (key == "cats" || key == "trees") return fv <= v
    if (key == "pomeranians" || key == "goldfish") return fv > v
    return fv == v
  }
  println("Part Two = ${Day16.solve(sues, foundAttrs, ::filter)}")
}
